src/crimson/osd/pg.h

```diff
@@ -573,7 +599,41 @@ class PG : public boost::intrusive_ref_counter<
   friend class PGAdvanceMap;
   friend class PeeringEvent;
   friend class RepRequest;
-};
+public:
```

Comment:
instead of putting the backfill related code right into `PG`, can we have a dedicated class for it, and plug that class into `PG`? like how we implement heartbeat in OSD?

Reply:
I'm still working on this one

Reply:
Um, it seems that, as data in OSD is stored in PGs, the recovery has to be done in terms of PGs as well, and the recovery involves not only PG but also other components like "background_recovery" operation or pull/push operations, so I think whether we put the recovery/backfill code in a dedicated class or not, we still have to keep PG's recovery/backfill related interfaces, which means the OSD's heartbeat approach may not be feasible. Am I right?

Reply:
i think what i am after is a way to decouple the related but relatively orthogonal functions into different classes. currently, PG is responsible for multiple things 

- I/O
- peering
- recovering
- backfill
- scrub

they are connected and in classic OSD, they are implemented in a single blob. but i think we are trying to extract the reusable logic from classic OSD, so that they can be shared by crimson and classic OSD. PeeringState is one of them. i think, from crimson's perspective, the beauty of this approach is more maintainable code, and less man-hour to develop the functionality.

in this case, i am more interested in more readable code. if we can extract the recovery out of PG, and let PG just expose the necessary facilities like acting set, pg log, state transition for the recovery, it'd be simpler to implement recovery and it'd be very helpful to test this feature.

> so I think whether we put the recovery/backfill code in a dedicated class or not, we still have to keep PG's recovery/backfill related interfaces, which means the OSD's heartbeat approach may not be feasible. Am I right?

yes and no. the related interface will be called only by the callers interested in recovery. heartbeat of OSD is relatively isolated in comparison with recovery in PG, but we can use a more sophisticated approach to compose them. one way i can think of is to design PG as a service which expose the facilities to its subcomponents, like scrub and recovery. and scrub and recovery will use these facilities to implement their features. 

these subcomponent could be designed as a member variables like heartbeat, or they can be implemented as mixed-in classes using CRTP pattern. see https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern

Reply:
Got your point:-)